File: code/mc_pi.py
import time
import sys
import numpy as np
from numpy.random import default_rng


def pi_iteration(circle_points: int):
	rng = default_rng()

	x = rng.uniform(-1.0, 1.0)
	y = rng.uniform(-1.0, 1.0)
	
	if x ** 2 + y ** 2 <= 1:
		circle_points += 1
	
	return circle_points


def pi_estimate(desired_iterations):
	"""
	Simple method of estimating PI by comparing randomly sampled point
	inside the area of a circle and a square (of diameter/length 1).
	
	:param desired_iterations: INT,
		Number of desired iterations to estimate PI
	:return: FLOAT,
		Estimate of PI
	"""
	# INITIALIZING NECESSARY PARAMETERS
	# It is notable that ALL points will be within the square, so we can
	# initially set the number of iterations as the variable
	# part_of_square
	increment = 0
	part_of_circle = 0
	part_of_square = desired_iterations
	
	while increment < desired_iterations:
		part_of_circle = pi_iteration(part_of_circle)
		increment += 1
	
	return 4 * part_of_circle / part_of_square


def main():
	if len(sys.argv) > 1: 
		iterations = float(sys.argv[1])
		print_flag = int(sys.argv[2])
	else: 
		iterations = 1e5
		print_flag = 0
	
	start = time.time()
	mc_pi = pi_estimate(iterations)
	end = time.time()
	
	if print_flag == 1:
		print(f"\nAfter {iterations:.0e} iterations\n"
		      f"\nEstimation of PI:\t PI = {mc_pi}"
		      f"\nCompare to NumPy:\t PI = {np.pi}"
		      f"\nFirst 10 digits:\t PI = 3.141592653 ...\n"
		      f"\nExecution time was {end-start:1.5e}s.\n")
	elif print_flag == 2:
		print(f"{end-start:1.5e}")


if __name__ == "__main__":
	main()

File: code/mc_pi.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#define RMAX ((double)RAND_MAX*(double)RAND_MAX)


int main(int argc, char* argv[])
{
	int i, print_flag = 0;
	double rand_x, rand_y, origin_dist, pi;
	double interval = 1e5;
	double circle_points = 0., square_points = 0.;
	clock_t start, end;
	double elapsed;
	
	// START A TIMER FOR CODE EXECUTION
	start = clock();
	
	// USE COMMAND LINE INPUT AS ITERATION LENGTH
	if (argc > 1) 
	{
		interval =  atof(argv[1]);
		print_flag = atoi(argv[2]);
	}
	
	// LOOP FOR SPECIFIED NUMBER OF ITERATIONS
	for (i = 0; i <= interval; i++)
	{
		// GENERATE A RANDOM NUMBER FROM A UNIFORM DISTRIBUTION
		// BETWEEN (-1, 1).
		rand_x = rand();
		rand_y = rand();
		
		// CALCULATE DISTANCE TO ORIGIN
		origin_dist = rand_x * rand_x + rand_y * rand_y;
		
		// CHECK IF DISTANCE TO THE ORIGIN IS WITHIN THE CIRCLE
		if (origin_dist <= RMAX) circle_points++;
		
		// INCREMENT NUMBER OF POINTS IN SQUARE
		square_points++;
	}
	
	// ESTIMATE PI AFTER END OF ITERATIONS
	pi = (4 * circle_points) / square_points;
	
	// NOTE END TIME OF EXECUTION AND CALCULATE DURATION
	end = clock();
	elapsed = ((double)(end - start)) / CLOCKS_PER_SEC;
	
	// PRINT FINAL VALUES (IF PROMPTED)
	if (print_flag==1) 
	{
		printf("\nAfter %.0e Iterations\n", interval);
		printf("\nEstimation of PI:\t PI = %f", pi);
		printf("\nCompare to MATH:\t PI = %f", M_PI);
		printf("\nFirst 10 digits:\t PI = 3.141592653 ...\n");
		printf("\nExecution time was %1.5es.\n\n", elapsed);
	}
	
	if (print_flag==2)
	{
		printf("%1.5e", elapsed);
	}

    return 0;
}
